src/main/cpp/webp_codec.cpp

```diff
@@ -0,0 +1,104 @@
+#include "webp_codec.h"
+#include "webp/decode.h"
+#include "webp/encode.h"
+#include <cstdint>
+#include <stdexcept>
+
+/*
+ * WEBPEncoder
+ */
+
+libench::WEBPEncoder::WEBPEncoder() {
+  WebPMemoryWriterInit(&this->writer_);
+};
+
+libench::WEBPEncoder::~WEBPEncoder() {
+  WebPMemoryWriterClear(&this->writer_);
+};
+
+libench::CodestreamContext libench::WEBPEncoder::encodeRGB8(const ImageContext &image) {
+  return this->encode8(image);
+}
+
+libench::CodestreamContext libench::WEBPEncoder::encodeRGBA8(const ImageContext &image) {
+  return this->encode8(image);
+}
+
+libench::CodestreamContext libench::WEBPEncoder::encode8(const ImageContext &image) {
+  WebPMemoryWriterClear(&this->writer_);
+
+  WebPConfig config;
+  WebPPicture pic;
+  const int level = 6;  // 0 (faster) - 9 (slower)
+  if (!WebPConfigInit(&config) || !WebPConfigLosslessPreset(&config, level) || !WebPPictureInit(&pic))
+    throw std::runtime_error("WEBP encode failed");
+
+  const int num_comps = image.format.comps.num_comps;
+  const int rgb_stride = image.width * num_comps;
+  pic.width = image.width;
+  pic.height = image.height;
+  pic.use_argb = 1;
+  int ret = num_comps == 3 ?
+      WebPPictureImportRGB(&pic, image.planes8[0], rgb_stride) :
+      WebPPictureImportRGBA(&pic, image.planes8[0], rgb_stride);
+  if (!ret) {
+    WebPPictureFree(&pic);
+    throw std::runtime_error("WEBP encode failed");
+  }
+
+  // Retain pixel values in fully transparent areas. The default will modify
+  // the (invisible) pixels to improve compression.
+  config.exact = 1;
```

Comment:
If using the default of 0 is acceptable, all of this code could be replaced with a single call similar to the decode:

https://developers.google.com/speed/webp/docs/api#simple_encoding_api